Attempt to fix plugin deletion error under Windows.

mpyat2 opened the pull request:
Issue #7: plug-ins cannot be deleted under Windows.
The fix attempts to resolve the issue: ClassLoader-s and streams are closed while plugin management.
Also, a small piece of code has been added to MenuBar.java to prevent starting PluginManagementDialog twice (currently this is possible even if PluginManagementDialog were modal: there is enough time to click the menu many times while initialization of PluginManagementDialog).
After any plugin-management-related operation, a user prompted to restart VStar.

dbenn commented on src/org/aavso/tools/vstar/ui/MenuBar.java:
```diff
@@ -1199,18 +1199,40 @@ public ActionListener createPluginManagerListener() {
 		return new ActionListener() {
 			@Override
 			public void actionPerformed(ActionEvent e) {
+				// Max: probably I'm paranoid: preventing opening PluginManagementDialog twice.
```
A reasonable concern in fact. Thanks, well spotted.

dbenn commented on src/org/aavso/tools/vstar/ui/MenuBar.java:
```diff
@@ -1199,18 +1199,40 @@ public ActionListener createPluginManagerListener() {
 			public void actionPerformed(ActionEvent e) {
+				// Max: probably I'm paranoid: preventing opening PluginManagementDialog twice.
+				if (PluginManagementDialog.getPluginManagementActive()) return;
+				PluginManagementDialog.setPluginManagementActive(true);
```
It's probably best to set this after line 1233 when the plugin manager has been fully activated. Indeed, one possible scenario is that the user clicks multiple times, causing this method to be invoked more than once, such that execution of a thread may be suspended after the active setter on line 1204 while a second enters the method, going no further than 1203 because the flag has already been set. The first thread would still resume to execute the rest and invoke the manager though, so probably okay. Adding the synchronised keyword to the method would allow the whole method to be treated as a critical section. That should work well with the active flag setting at line 1233. Am I making sense? It's later here. :)

mpyat2 replied:
I think no. This code is intended to prevent the secondary execution of the menu action so it should be at the top of the "actionPerformed". The same check is done in PluginSettingsPane.java, the second place where one of the plugin management operations (deleteAllPlugins) is executed. "pluginManagementActive" flag can be placed elsewhere, PluginManagementDialog is not the unique place. pluginManagementActive flag is normally accessed from the main thread except one place: an exception handler at the line 1221 of MenuBar.jar where the flag is reset. This is why I declared it as "volatile" initially, yet synchronized getter and setter probably even better.

dbenn commented:
I have no easy access to a Windows laptop at home. I'm guessing you've tested this on Windows @mpyat2

clkotnik commented:
I built this branch on my Windows 10 PC and tested it.  I deleted individual plugins, added individual plugins and exercised the delete all plugins from the prefs dialog.  In all cases it behaved as expected.

mpyat2 commented:
Hi Cliff,

Thank you!

Cheers,
Max

On Thu, 23 Jul 2020, 18:03 clkotnik, <notifications@github.com> wrote:

> I built this branch on my Windows 10 PC and tested it. I deleted
> individual plugins, added individual plugins and exercised the delete all
> plugins from the prefs dialog. In all cases it behaved as expected.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AAVSO/VStar/pull/33#issuecomment-663058704>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWWZCIDRV2ILK3UN2ICWCTR5BGK3ANCNFSM4OAS66IA>
> .
>

dbenn commented:
@mpyat2 @clkotnik Hi Max and Cliff. I've carried out those tests on the branch under Mac OS X to make sure it still works. All good!

Thanks guys. 

Max, these changes will be appreciated by Brad's course participants in August, and all Windows users. Appreciate it!

I'll merge the branch now.
